File: packages/lynx-webrtc/android/com/sigx/webrtc/WebRTCPublisher.kt
package com.sigx.webrtc

import android.util.Log
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.tasm.LynxView
import java.util.UUID

/**
 * Per-[LynxView] publisher that pumps [WebRTCEventBus] payloads into JS
 * via `LynxView.sendGlobalEvent("__sigxWebRTCEvent", [...])`.
 *
 * One instance per LynxView; instantiated by the generated
 * `GeneratedLifecyclePublishers.attachAll(lynxView)` and retained for the
 * LynxView's lifetime.
 */
class WebRTCPublisher(private val lynxView: LynxView) {

    private var token: UUID? = null

    fun attach() {
        token = WebRTCEventBus.addListener { payload ->
            try {
                val params = JavaOnlyArray()
                params.pushMap(payload)
                lynxView.sendGlobalEvent(EVENT_NAME, params)
            } catch (e: Throwable) {
                Log.w(TAG, "publish failed: ${e.message}")
            }
        }
    }

    fun detach() {
        token?.let { WebRTCEventBus.removeListener(it) }
        token = null
    }

    private companion object {
        const val TAG = "SigxWebRTCPublisher"
        const val EVENT_NAME = "__sigxWebRTCEvent"
    }
}

File: packages/lynx-webrtc/ios/WebRTCPublisher.swift
import Foundation
import Lynx

/// Per-`LynxView` publisher that pumps `WebRTCEventBus` payloads into JS
/// via `LynxView.sendGlobalEvent("__sigxWebRTCEvent", [...])`.
///
/// One instance per LynxView; instantiated by the generated
/// `GeneratedLifecyclePublishers.attachAll(to:)` and retained for the
/// LynxView's lifetime.
final class WebRTCPublisher {

    private weak var lynxView: LynxView?
    private var token: UUID?

    init(lynxView: LynxView) {
        self.lynxView = lynxView
        self.token = WebRTCEventBus.shared.addListener { [weak self] json in
            guard let view = self?.lynxView else { return }
            // Pass a single JSON string as the only param — the JS shim
            // parses string events. A string survives Lynx 0.5.0's bridge
            // marshalling intact where a structured map carrying a nested
            // map (e.g. `candidate`) does not (#342).
            view.sendGlobalEvent("__sigxWebRTCEvent", withParams: [json])
        }
    }

    deinit {
        if let token = token {
            WebRTCEventBus.shared.removeListener(token)
        }
    }
}
